src/rust_option.rs

```diff
@@ -0,0 +1,266 @@
+#![allow(missing_docs)]
+
+#[cfg(feature = "alloc")]
+use alloc::boxed::Box;
+use core::mem::ManuallyDrop;
+use core::ops::Deref;
+use core::pin::Pin;
+
+mod private {
+    pub trait Sealed {}
+}
+pub trait OptionType: private::Sealed {}
+
+impl<T: Sized> private::Sealed for &T {}
+impl<T: Sized> OptionType for &T {}
+
+impl<T: Sized> private::Sealed for &mut T {}
+impl<T: Sized> OptionType for &mut T {}
+
+impl<T: Sized> private::Sealed for Pin<&mut T> {}
+impl<T: Sized> OptionType for Pin<&mut T> {}
+
+impl<T: Sized> private::Sealed for *const T {}
+impl<T: Sized> OptionType for *const T {}
+
+impl<T: Sized> private::Sealed for *mut T {}
+impl<T: Sized> OptionType for *mut T {}
+
+#[cfg(feature = "alloc")]
+impl<T: Sized> private::Sealed for Box<T> {}
+#[cfg(feature = "alloc")]
+impl<T: Sized> OptionType for Box<T> {}
+
+#[repr(C)]
+union OptionInner<T: OptionType> {
+    value: ManuallyDrop<T>,
+    empty: usize,
+}
+
+// ABI compatible with C++ rust::Option<T> (not necessarily core::option::Option<T>).
+#[repr(C)]
+pub struct RustOption<T: OptionType> {
```

Comment:
we need the bound on the struct itself (rather than just the impls) because we need it for the Drop impl